src/runtime/objmodel.cpp

```diff
@@ -3804,7 +3802,85 @@ extern "C" Box* unaryop(Box* operand, int op_type) {
     return rtn;
 }
 
-extern "C" Box* getitem(Box* value, Box* slice) {
+Box* callItemAttr(Box* target, const std::string* item_str, Box* item, Box* value, CallRewriteArgs* rewrite_args) {
+    if (value) {
+        return callattrInternal2(target, item_str, CLASS_ONLY, rewrite_args, ArgPassSpec(2), item, value);
+    } else {
+        return callattrInternal1(target, item_str, CLASS_ONLY, rewrite_args, ArgPassSpec(1), item);
+    }
+}
+
+// This function decides whether to call the slice operator (e.g. __getslice__)
+// or the item operator (__getitem__).
+Box* callItemOrSliceAttr(Box* target, const std::string* item_str, const std::string* slice_str, Box* slice, Box* value,
+                         CallRewriteArgs* rewrite_args) {
+    // If we don't have a slice operator, fall back to item operator.
+    // No rewriting for the typeLookup here (i.e. no guards if there is no slice operator). These guards
+    // will be added when we call callattrInternal on a slice operator later. If the guards fail, we fall
+    // back into the slow path which has this fallback to the item operator.
+    Box* slice_attr = typeLookup(target->cls, *slice_str, NULL);
```

Comment:
I think this typeLookup call needs to use rewrite args so we can emit guards that the type has that attribute (and it has the value) we think it does